File: cpp/endpoints/docking/delete_dock.cpp
#include "rclcpp/rclcpp.hpp"
#include "clearpath_dock_msgs/srv/remove_dock.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <yaml-cpp/yaml.h>

using namespace std::chrono_literals;

// TODO: Enter the name of the dock to be removed
const std::string DOCK_NAME = "";

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
  std::string namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("delete_dock_client");
  rclcpp::Client<clearpath_dock_msgs::srv::RemoveDock>::SharedPtr client =
    node->create_client<clearpath_dock_msgs::srv::RemoveDock>("/" + namespace_ + "/" + std::string("docking/dock_manager/delete_dock"));

  auto request = std::make_shared<clearpath_dock_msgs::srv::RemoveDock::Request>();
  request->name = DOCK_NAME;

  while (!client->wait_for_service(2s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("delete_dock_client"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("delete_dock_client"), "/%s/docking/dock_manager/delete_dock service not available, waiting again...", namespace_.c_str());
  }

  auto result = client->async_send_request(request);

  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    if (result.get()->success)
    {
      RCLCPP_INFO(rclcpp::get_logger("delete_dock_client"), "Dock deleted successfully!");
    }
    else
    {
      RCLCPP_ERROR(rclcpp::get_logger("delete_dock_client"), "Failed to delete dock!");
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("delete_dock_client"), "Failed to call service /%s/docking/dock_manager/delete_dock", namespace_.c_str());
  }

  rclcpp::shutdown();
  return 0;
}

File: python/endpoints/docking/delete_dock.py
from clearpath_config.common.utils.yaml import read_yaml
import rclpy
from rclpy.node import Node
from clearpath_dock_msgs.srv import RemoveDock

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'DeleteDock'

# TODO: Enter the name of the dock to be removed
DOCK_NAME: str = ''


class DeleteDock(Node):

    def __init__(self, namespace: str):
        super().__init__('delete_dock_client')
        self._delete_dock_client = self.create_client(RemoveDock, f'/{namespace}/docking/dock_manager/delete_dock')

        while not self._delete_dock_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().info(f'[{LOGGING_NAME}] /{namespace}/docking/dock_manager/delete_dock service not available, waiting again...')

    def delete_dock(self):
        req = RemoveDock.Request()
        req.name = DOCK_NAME
        self.future = self._delete_dock_client.call_async(req)
        self.future.add_done_callback(self.delete_dock_srv_response_cb)

    def delete_dock_srv_response_cb(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'[{LOGGING_NAME}] Dock deleted successfully!')
            else:
                self.get_logger().error(f'[{LOGGING_NAME}] Failed to delete dock: {response.message}')
        except Exception as e:
            self.get_logger().error(f'[{LOGGING_NAME}] Service call failed: {e}')

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    delete_dock_client = DeleteDock(namespace)
    delete_dock_client.delete_dock()

    rclpy.spin(delete_dock_client)


if __name__ == '__main__':
    main()
